kvserver: use engine only after entry cache

pav-kv opened the pull request:
Raft entries are often in the entries cache. This PR decouples `term` and `entries` fetching to prepare for the refactoring, and moves constructing a storage Reader strictly after checking the cache, to reduce the overhead.

Fixes #92274
Release note: None

pav-kv committed:
kvserver: use engine only after entry cache

Raft entries are often in the entries cache. This commit puts constructing a
storage Reader strictly after checking the cache, to reduce the overhead.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/92326)
<!-- Reviewable:end -->

pavelkalinnikov commented on pkg/kv/kvserver/replica_raftstorage.go:
```diff
@@ -284,16 +284,18 @@ func (r *Replica) GetTerm(i uint64) (uint64, error) {
+	ents, err := entries(ctx, rsl, eng, rangeID, eCache, nil /* sideloaded */, i, i+1, math.MaxUint64 /* maxBytes */)
 	if errors.Is(err, raft.ErrCompacted) {
-		ts, err := rsl.LoadRaftTruncatedState(ctx, reader)
+		// TODO(pavelkalinnikov): maybe push this check to entries(), because it
```
Seemingly, the only thing that `term` and `entries` do differently here is:
- `term` loads the truncated state unconditionally, because it would contain the term of the last compacted entry
- `entries` only loads it when it could not read any entries

tbg replied:
The code in `entries` seems a bit more complicated than it needs to be. On the face of it this could be explained by a desire to assert something. But really the assertions are quite weird; note how we maintain `expectedIndex` but (in `scanFunc`) silently bail out if we find a gap. I think *that* is the one place where we should fail if we see a gap. 
Then, when the scan is done, we should check

- if there was an entry, check that it matches `lo`. If so,
- check if we had the number of entries requested or we hit the max bytes limit. If so,
- we're good.
- Otherwise, load first and return ErrCompacted if lo < the first index.
- Otherwise, load last index and return ErrUnavailable if hi > lastIndex.
- otherwise, return an assertion failed error.

I agree that we should be able to do the code movement you suggested now. Perhaps we can clean up `entries` too as per the above, either in a separate commit, or PR.

pavelkalinnikov replied:
I decoupled `term` and `entries`, so this TODO is no longer a concern.

Will look at refactoring `entries` in a follow-up PR.

tbg commented on pkg/kv/kvserver/replica_raftstorage.go:
```diff
@@ -259,13 +259,13 @@ func (r *replicaRaftStorage) Term(i uint64) (uint64, error) {
 		return r.mu.lastTerm, nil
 	}
 	// Try to retrieve the term for the desired entry from the entry cache.
+	// TODO(pavelkalinnikov): maybe drop this, because term calls entries, which
```
Sounds reasonable. The difference between `cache.Get` and `cache.Scan` isn't big; Get internally looks like Scan.

pavelkalinnikov replied:
Done even a better thing, see the commits. I figured that decoupling `term` from `entries` solves this and a few other problems.

pav-kv committed:
kvserver: make maxBytes handling clearer

A faster fast path, and better readability (a little copying is better than a
little dependency - for the append).

Release note: None

pav-kv committed:
kvserver: rm unreachable condition

Both cache and non-cache part of the entries reading code stops if it sees a
missing index, and does not add entries coming afte the gap. We should not
return a "valid" ErrCompacted error here, but instead either return a "bad"
error, or not check this condition at all.

Release note: None

pav-kv committed:
kvserver: decouple term fetching from entries func

Previously the entries func was used both for fetching entries and term. These
were quite different:

- term is fetched for a single entry, and doesn't use sideloaded storage
- entries are fetched in batches, use sideloaded storage, and respect maxSize

Separating the two functions is good (despite a little bit of code duplication)
for a bunch of reasons:

- clearer readability / one thing at a time
- possibility to optimize separately because these are different workloads
- ... which is immediately done in this commit (use cache Get instead of Scan)

Release note: None

pavelkalinnikov commented:
@tbg This is now ready for a full review.

pavelkalinnikov commented on pkg/kv/kvserver/replica_raftstorage.go:
```diff
@@ -278,27 +262,55 @@ func term(
 	}
-	if len(ents) == 0 {
-		return 0, nil
+	if index == ts.Index {
```
Probably also need to return a "bad" error if index > ts.Index (and < lastIndex). This would mean that there is a gap.

pavelkalinnikov replied:
Done.

tbg reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r2, 1 of 1 files at r3, 2 of 2 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92326#-:-NI7O5kn5KyCAIVxfVvv:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker, @nvanbenschoten, and @pavelkalinnikov)

___
*[`pkg/kv/kvserver/replica_raftstorage.go` line 118 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/92326#-NI7LwzH3sYV_0-7E24G:-NI7LwzH3sYV_0-7E24H:bqa8cv8) ([raw file](https://github.com/cockroachdb/cockroach/blob/89c99ea0241ab7d3040c7ee4e7397697cc8ca9aa/pkg/kv/kvserver/replica_raftstorage.go#L118)):*
> ```Go
> // entries retrieves entries from the engine. It inlines the sideloaded entries,
> // and caches all the loaded entries. The size of the returned entries does not
> // exceed maxSize, unless only one entry is returned.
> ```

We should consider returning all of the entries we read anyway, i.e. interpret `maxSize` as a target size.
___
*[`pkg/kv/kvserver/replica_raftstorage.go` line 291 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/92326#-NI7MpGp7UJV4_kDfiT7:-NI7MpGp7UJV4_kDfiT8:bx2davl) ([raw file](https://github.com/cockroachdb/cockroach/blob/89c99ea0241ab7d3040c7ee4e7397697cc8ca9aa/pkg/kv/kvserver/replica_raftstorage.go#L291)):*
> ```Go
> 		// sideloaded entries here to keep the term fetching cheap.
> 		if !logstore.SniffSideloadedRaftCommand(entry.Data) {
> 			eCache.Add(rangeID, []raftpb.Entry{entry}, false /* truncate */)
> ```

I suspect it makes little sense to populate the cache here but let's keep it for parity with the old code until we can measure the impact of changes.
___
*[`pkg/kv/kvserver/replica_raftstorage.go` line 296 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/92326#-NI7NI612vWQXmzdahpS:-NI7NI612vWQXmzdahpT:bql235n) ([raw file](https://github.com/cockroachdb/cockroach/blob/89c99ea0241ab7d3040c7ee4e7397697cc8ca9aa/pkg/kv/kvserver/replica_raftstorage.go#L296)):*
> ```Go
> 	}
> 	// Otherwise, the entry at the given index is not found. This can happen if
> 	// the index is beyond the current log size, or it has been compacted away.
> ```

nit "log size" is misleading here, better "if the index is ahead of the lastIndex, or it has been ..."


<!-- Sent from Reviewable.io -->

pav-kv reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92326)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @erikgrinaker, @nvanbenschoten, and @tbg)

___
*[`pkg/kv/kvserver/replica_raftstorage.go` line 118 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/92326#-NI7LwzH3sYV_0-7E24G:-NI7a8GgCIkT-PmoMPEr:b-7yd1q4) ([raw file](https://github.com/cockroachdb/cockroach/blob/89c99ea0241ab7d3040c7ee4e7397697cc8ca9aa/pkg/kv/kvserver/replica_raftstorage.go#L118)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

We should consider returning all of the entries we read anyway, i.e. interpret `maxSize` as a target size.
</blockquote></details>

Added a TODO.
___
*[`pkg/kv/kvserver/replica_raftstorage.go` line 291 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/92326#-NI7MpGp7UJV4_kDfiT7:-NI7aC-o23Gcu58SFX9n:b-7yd1q4) ([raw file](https://github.com/cockroachdb/cockroach/blob/89c99ea0241ab7d3040c7ee4e7397697cc8ca9aa/pkg/kv/kvserver/replica_raftstorage.go#L291)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I suspect it makes little sense to populate the cache here but let's keep it for parity with the old code until we can measure the impact of changes.
</blockquote></details>

Added a TODO.
___
*[`pkg/kv/kvserver/replica_raftstorage.go` line 296 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/92326#-NI7NI612vWQXmzdahpS:-NI7aEz4DozxsqRW5BUB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/89c99ea0241ab7d3040c7ee4e7397697cc8ca9aa/pkg/kv/kvserver/replica_raftstorage.go#L296)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

nit "log size" is misleading here, better "if the index is ahead of the lastIndex, or it has been ..."
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

pavelkalinnikov commented:
bors r=tbg
